### ISSUE TITLE
Why do I encounter "Illegal instruction" error?

### ISSUE THREAD

**Issue by epioneMW** [Illumina/Nirvana#23]
I built the app as instructed and tried to run dotnet Nirvana/2.0.7/bin/Release/netcoreapp2.0/Nirvana.dll --in HiSeq.10000.vcf --cache Data/Cache/26/GRCh37/Ensembl84  --sd Data/SupplementaryDatabase/44/GRCh37 --ref Data/References/5/Homo_sapiens.GRCh37.Nirvana.dat  --out HiSeq.10000.annotated.

But I am having error saying "Illegal instruction". How can I solve this problem?

**Comment by ivanwick** [comment]
Nirvana loads unmanaged native code for its fast data compression routines. The bundled BlockCompression library was built with `-arch:AVX` so it uses [AVX CPU instructions](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions).

I am also getting `SIGILL, Illegal instruction` when running on an older CPU that doesn't support AVX. Intel(R) Core(TM)2 Duo CPU     P8800  @ 2.66GHz.

Stack trace when the illegal instruction happens:
```
#0  0x00007ffff13f4ed3 in HUF_readDTableX4 () from /root/nirvana/2.0.9/libBlockCompression.so
#1  0x00007ffff13f4aa8 in HUF_decompress4X4_DCtx () from /root/nirvana/2.0.9/libBlockCompression.so
#2  0x00007ffff13f710e in ZSTD_decodeLiteralsBlock () from /root/nirvana/2.0.9/libBlockCompression.so
#3  0x00007ffff13f73b6 in ZSTD_decompressBlock_internal () from /root/nirvana/2.0.9/libBlockCompression.so
#4  0x00007ffff13f8167 in ZSTD_decompressFrame () from /root/nirvana/2.0.9/libBlockCompression.so
#5  0x00007ffff13f82f4 in ZSTD_decompressDCtx () from /root/nirvana/2.0.9/libBlockCompression.so
#6  0x00007ffff13f8342 in ZSTD_decompress () from /root/nirvana/2.0.9/libBlockCompression.so
#7  0x00007fff7d106ee1 in ?? ()
#8  0x00000000001c8ea9 in ?? ()
#9  0x00007ffff6808c78 in vtable for InlinedCallFrame () from /usr/share/dotnet/shared/Microsoft.NETCore.App/2.2.0/libcoreclr.so
#10 0x00007fffffffc8d8 in ?? ()
#11 0x00007fff7d65cc38 in ?? ()
#12 0x00007fff7d65cc38 in ?? ()
#13 0x00007fffffffc0c0 in ?? ()
#14 0x00007fff7d106ee1 in ?? ()
#15 0x00007fffffffc140 in ?? ()
#16 0x00007fff7d65cc38 in ?? ()
#17 0x00007fff71008d40 in ?? ()
#18 0x00007fff70008d08 in ?? ()
#19 0x00007fff600c4ee8 in ?? ()
#20 0x0000000000000000 in ?? ()
```

The illegal instruction is
```
=> 0x00007ffff13f4ed3 <HUF_readDTableX4+659>:   vmovdqu (%r8),%xmm2
```
which is indeed an AVX instruction in Nirvana's bundled build of ZSTD. facebook/zstd#1183

What CPU are you running it on? If it is older and doesn't support AVX, then that could the cause.

I don't see any tooling in the git repository to build libBlockCompression.so from source, it is only available as a packaged binary. It would be nice to be able to customize the arch flags for compatibility with different CPUs.

**Comment by MichaelStromberg** [comment]
Sorry about that. Originally, we were overly eager to compile the BlockCompression library with AVX support. We ended up back-pedaling on this since we found enough machines that didn't have those CPU instructions. If you ever want to roll your own BlockCompression library for a new platform or if you want to tweak it for a specific CPU architecture, here's the public git repo: https://github.com/Illumina/BlockCompression
